### DISCUSSION TITLE
[Feature] auto select best featured photo

### DISCUSSION THREAD

**Discussion by 1Euro7Cent** [immich-app/immich#10051] [Feature Request]
### I have searched the existing feature requests to make sure this is not a duplicate request.

- [X] Yes

### The feature

I have some person's faces recognized and immich auto selects a featured photo. After some time a better one is processed but the first one stays.

The request:
if the featured photo is NOT manually set search for new best face photo. 
if a better one is found, set as featured 

Here is an example. The big photo (left) is much better than the selected one (right)
![image](https://github.com/immich-app/immich/assets/67194495/76d44ac6-749a-4768-a9e0-8ef4e65afe36)

I know that I can just set it. But it would be a QoL feature.

### Platform

- [X] Server
- [ ] Web
- [ ] Mobile

**Comment by bo0tzz** [comment]
How do you imagine determining whether a photo is "better"?

**Reply by 1Euro7Cent** [reply]
One variable that could count for it is the size of the image. If it is 30px probably not good.

And the rest maybe another AI feature

**Reply by what-in-the-nim** [reply]
Here are some metrics I can think of

- Face resolution (pixel x pixel)
- Face orientation (yaw, pitch, roll) -> [link](https://github.com/jerryhouuu/Face-Yaw-Roll-Pitch-from-Pose-Estimation-using-OpenCV)
- Face location & size (e.g. biggest face among other faces (group photos) or face is centered (selfies))
- Bluriness (use variance of laplacian) -> [link](https://pyimagesearch.com/2015/09/07/blur-detection-with-opencv/)
- Colorful (choose colorful >> black and white) -> [link](https://pyimagesearch.com/2017/06/05/computing-image-colorfulness-with-opencv-and-python/)

To compute the "better" score, we might need to determine how much weight to put in each metrics (linear combination).

**Comment by kfoxton** [comment]
I would like this too. I find that during the import process of my existing photos, when the faces of people were originally selected, they may have been a background person, so the featured photo of them is very pixelated/blurry. Now that I have all of my photos imported there are many better selections that could be made, that the face of the person is primary in the photo, and would be clear as a featured photo.

While this can be done manually, when I have thousands of faces in my Immich setup, it would take a signifiantly long time to complete manually.

**Comment by gabbro246** [comment]
A better thumbnail could be chosen based on resolution, like @1Euro7Cent said, or based on sharpness (I believe there are efficient algorithms for that). Another approach could be using a value returned from face detection or recognition.

Generally, I think this is a well-explored problem. I would love such a feature since many faces in my instance are hidden due to depth of field, and an algorithm could detect them better than I can recognize them without opening the full photo.

**Comment by ireun** [comment]
Yes please!
Some ideas - after first grouping:
  - re-select a best photo possible as `main photo` for every person detected
    - best meaning: highest resolution, not blurred, etc
  - or at least let us do that afterward 
  - reasoning: I think currently it just selects first photo found as 'main' and it usually is not best at all

**Reply by ireun** [reply]
Or maybe face-aligmnment ML? Then select a high resolution & straight-on photo for a person

**Comment by FraudR** [comment]
I have a bunch of compressed images from messaging apps that end up being the featured photo. Sorting the faces by resolution and picking the highest one would be a great feature. Below is an example of the problem.

<img width="168" height="212" alt="image" src="https://github.com/user-attachments/assets/39bae0dd-69db-43a3-a885-cc5cc55f8f3f" />

**Comment by GreaseHeadD** [comment]
I thought this would be a good programming exercise since I have a lot of people with blurry photos too, so I created a python script to fix it. I incorporated @1Euro7Cent idea of using the size of the image as a score, using the facial recognition bounding box. I also added the possibility to let it determine the most sharpest image as @gabbro246 suggested.

I also added some more options like setting featured photos based on the person's birthday (if it's set), adding a recency score and the ability to specific a certain date and range for photos to select. The script can update all people and also specific people using ids, names and if they are favorite.

I've put the script here: [https://github.com/GreaseHeadD/immich-featured-photo-organizer](https://github.com/GreaseHeadD/immich-featured-photo-organizer)
And I have tried to test and document it as thoroughly as I could, but let me know if something's wrong.

**Comment by ireun** [comment]
This is probably related to https://github.com/immich-app/immich/discussions/6348

**Reply by 1Euro7Cent** [reply]
Yes. But this one seems to have a working version from the community and more contributions than https://github.com/immich-app/immich/discussions/6348
